### PR TITLE
shell: Navigate to new host after adding one

### PR DESCRIPTION
**Pull request by mvollmer**
- [ ] #20826

### CODE REVIEW

File: pkg/shell/hosts.jsx [hunk 1]
```diff
@@ -89,6 +89,16 @@ export class CockpitHosts extends React.Component {
...
+
+        window.trigger_connection_flow = machine => {
+            if (!this.state.modal_properties)
+                this.connectHost(machine);
+        };
+        this.props.index.navigate(null, true);
+    }
+
+    componentWillUnmount() {
+        window.trigger_connection_flow = null;
```
Comment (cockpituous) [inline pkg/shell/hosts.jsx:101]: These 2 added lines are not executed by any test.

File: pkg/shell/hosts.jsx [hunk 2]
```diff
@@ -116,12 +126,86 @@
...
+    }
+
+    async onHostEdit(event, machine) {
+        const connection_string = await this.showModal({ address: machine.address });
+        if (connection_string) {
+            const parts = this.props.machines.split_connection_string(connection_string);
+            const addr = this.props.hostAddr({ host: parts.address }, true);
+            if (machine == this.props.machine && parts.address != machine.address) {
+                this.props.loader.connect(parts.address);
+                this.props.jump(addr);
```
Comment (cockpituous) [inline pkg/shell/hosts.jsx:154]: These 3 added lines are not executed by any test.

File: pkg/shell/hosts.jsx [hunk 3]
```diff
@@ -116,12 +126,86 @@
...
+            connection_string = await this.showModal({
+                address: machine.address,
+                template: "connect"
+            });
+        } else {
+            // Try to connect without any dialog
+            try {
+                await try2Connect(this.props.machines, machine.connection_string);
+                connection_string = machine.connection_string;
+            } catch (err) {
```
Comment (cockpituous) [inline pkg/shell/hosts.jsx:182]: This added line is not executed by any test.

File: pkg/shell/hosts.jsx [hunk 4]
```diff
@@ -116,12 +126,86 @@
...
+                // continue with troubleshooting in the dialog
+                connection_string = await this.showModal({
+                    address: machine.address,
+                    template: codes[err.problem] || "change-port",
```
Comment (cockpituous) [inline pkg/shell/hosts.jsx:186]: This added line is not executed by any test.

File: pkg/shell/hosts_dialog.jsx [hunk 5]
```diff
@@ -124,6 +196,8 @@ class AddMachine extends React.Component {
             old_machine = props.machines_ins.lookup(props.old_address);
         if (old_machine)
             color = this.rgb2Hex(old_machine.color);
+        if (old_machine && !old_machine.visible)
+            old_machine = null;
```
Comment (cockpituous) [inline pkg/shell/hosts_dialog.jsx:200]: This added line is not executed by any test.

File: pkg/shell/hosts_dialog.jsx [hunk 6]
```diff
@@ -222,22 +296,27 @@
...
+        if (!window.sessionStorage.getItem("connection-warning-shown")) {
+            this.props.setError({ problem: "danger", command: "close" });
+        } else {
+            this.props.run(this.props.try2Connect(address), ex => {
+                if (ex.problem === "no-host") {
+                    let host_id_port = address;
+                    let port = "22";
+                    const port_index = host_id_port.lastIndexOf(":");
+                    if (port_index === -1) {
+                        host_id_port = address + ":22";
```
Comment (cockpituous) [inline pkg/shell/hosts_dialog.jsx:308]: These 2 added lines are not executed by any test.

File: pkg/shell/hosts_dialog.jsx [hunk 7]
```diff
@@ -902,6 +980,31 @@
...
+    return new Promise((resolve, reject) => {
+        const conn_options = { ...options, payload: "echo", host: address };
+
+        conn_options["init-superuser"] = get_init_superuser_for_options(conn_options);
+
+        const machine = machines_ins.lookup(address);
+        if (machine && machine.host_key && !machine.on_disk) {
+            conn_options['temp-session'] = false; // Compatibility option
+            conn_options.session = 'shared';
+            conn_options['host-key'] = machine.host_key;
```
Comment (cockpituous) [inline pkg/shell/hosts_dialog.jsx:993]: These 4 added lines are not executed by any test.

File: pkg/shell/indexes.jsx [hunk 8]
```diff
@@ -105,11 +102,31 @@ function MachinesIndex(index_options, machines, loader) {
     if (meta_multihost instanceof HTMLMetaElement && meta_multihost.content == "yes")
         host_switcher_enabled = true;
 
+    /* Should show warning before connecting? */
+    let config_ready = false;
+    cockpit.dbus(null, { bus: "internal" }).call("/config", "cockpit.Config", "GetString",
+                                                 ["Session", "WarnBeforeConnecting"], [])
+            .then(([result]) => {
+                if (result == "false" || result == "no") {
+                    window.sessionStorage.setItem("connection-warning-shown", "yes");
```
Comment (cockpituous) [inline pkg/shell/indexes.jsx:111]: These 3 added lines are not executed by any test.

**Review by garrett — CHANGES_REQUESTED** [review]
This looks _mostly_ great. Thanks!

However: After I added a new host, I got the connect warning dialog and then the key dialog, all as expected. However, it didn't switch to that remote host. Instead, it stayed on the original host and the switcher persisted. 

I would have expected it to switch to the new host and the switcher to go away at the same time.

(I thought I spotted some regressions, like the host switcher not having a hover state, but it looks like that in main too, so I guess we picked up that regression elsewhere, in the past. I guess after we land the rework and reimplement the shell's UI, it'll be automatically fixed. Otherwise, I could send up a PR to fix the CSS outside of your changes.)

**Comment by mvollmer** [comment]
Now part of #20826.

**Pull request closed without merging**